core/src/main/java/io/grpc/internal/ManagedChannelOrphanWrapper.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2018, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.ManagedChannel;
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.SoftReference;
+import java.lang.ref.WeakReference;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+import java.util.logging.Logger;
+
+class ManagedChannelOrphanWrapper extends ForwardingManagedChannel {
+  private static final ReferenceQueue<ManagedChannelOrphanWrapper> REFQUEUE =
+      new ReferenceQueue<ManagedChannelOrphanWrapper>();
+  // Retain the References so they don't get GC'd
+  private static final ConcurrentMap<ManagedChannelReference, ManagedChannelReference> REFS =
+      new ConcurrentHashMap<ManagedChannelReference, ManagedChannelReference>();
+  static final Logger logger = Logger.getLogger(ManagedChannelOrphanWrapper.class.getName());
+
+  private final ManagedChannelReference phantom;
+
+  ManagedChannelOrphanWrapper(OrphanWrappableManagedChannel delegate) {
+    this(delegate, REFQUEUE, REFS);
+  }
+
+  @VisibleForTesting
+  ManagedChannelOrphanWrapper(
+      OrphanWrappableManagedChannel delegate,
+      ReferenceQueue<ManagedChannelOrphanWrapper> refqueue,
+      ConcurrentMap<ManagedChannelReference, ManagedChannelReference> refs) {
+    super(delegate);
+    phantom = new ManagedChannelReference(this, delegate, refqueue, refs);
+    // zpencer: Can this be implemented with notifyWhenStateChanged and fire at SHUTDOWN instead?
+    // The reason OrphanableMC exists is because ManagedChannelImpl is both unmockable and
+    // difficult to construct.
+    delegate.setPhantom(phantom);
```

Comment:
Let's not do this. Let's either drop that part of the feature or use `awaitTermination` as a signal.

Reply:
`awaitTermination` in the forwarding class is a nice way to do this, we can remove the interface as well.